File: crates/solidity/outputs/npm/tests/src/doc-examples/simple-contract.ts
import { Language } from "@nomicfoundation/slang/language";
import { RuleNode, TokenNode } from "@nomicfoundation/slang/cst";
import { ProductionKind, RuleKind, TokenKind } from "@nomicfoundation/slang/kinds";

test("simple contract", () => {
  const language = new Language("0.8.0");
  const parseOutput = language.parse(ProductionKind.ContractDefinition, "contract Foo {}");

  const parseTree = parseOutput.parseTree as RuleNode;
  expect(parseTree.kind).toEqual(RuleKind.ContractDefinition);

  const children = parseTree.children;
  expect(children).toHaveLength(6);

  expect((children[0] as TokenNode).kind).toEqual(TokenKind.ContractKeyword);
  expect((children[1] as RuleNode).kind).toEqual(RuleKind.LeadingTrivia);
  expect((children[2] as TokenNode).kind).toEqual(TokenKind.Identifier);
  expect((children[3] as RuleNode).kind).toEqual(RuleKind.LeadingTrivia);
  expect((children[4] as TokenNode).kind).toEqual(TokenKind.OpenBrace);
  expect((children[5] as TokenNode).kind).toEqual(TokenKind.CloseBrace);
});

File: crates/solidity/outputs/cargo/tests/src/doc_examples/simple_contract.rs
use anyhow::Result;
use semver::Version;
use slang_solidity::{
    cst::Node,
    kinds::{RuleKind, TokenKind},
    language::Language,
};

#[test]
fn simple_contract() -> Result<()> {
    let language = Language::new(Version::parse("0.8.0")?)?;
    let parse_output = language.parse(RuleKind::ContractDefinition, "contract Foo {}");

    let parse_tree = parse_output.tree();
    let rule = parse_tree
        .into_rule()
        .expect("Expected root node to be a rule");

    assert_eq!(rule.kind, RuleKind::ContractDefinition);
    assert_eq!(rule.children.len(), 6);

    let children = &rule.children;
    assert!(
        matches!(&children[0], (_, Node::Token(token)) if token.kind == TokenKind::ContractKeyword)
    );
    assert!(matches!(&children[1], (_, Node::Rule(rule)) if rule.kind == RuleKind::LeadingTrivia));
    assert!(matches!(&children[2], (_, Node::Token(token)) if token.kind == TokenKind::Identifier));
    assert!(matches!(&children[3], (_, Node::Rule(rule)) if rule.kind == RuleKind::LeadingTrivia));
    assert!(matches!(&children[4], (_, Node::Token(token)) if token.kind == TokenKind::OpenBrace));
    assert!(matches!(&children[5], (_, Node::Token(token)) if token.kind == TokenKind::CloseBrace));

    assert_eq!(rule.unparse(), "contract Foo {}");

    Ok(())
}
